src/ddsrt/src/strtod.c

```diff
@@ -148,7 +148,7 @@ ddsrt_strtod(const char *nptr, char **endptr, double *dblptr)
     }
   }
 
-  if ((dbl == HUGE_VALF || dbl == HUGE_VALL || dbl == 0) && errno == ERANGE) {
+  if ((dbl == HUGE_VAL || dbl == 0) && errno == ERANGE) {
```

Comment:
I agree that `HUGE_VALF` and `HUGE_VALL` must have been put in because of a misreading of the spec, so removing those is definitely an imrpovement. There are other things amiss, though ...

Firstly, another misreading is the sign, because it states:
> 7.20.1.3:
> (¶10) If the correct value is outside the range of representable values, plus or minus `HUGE_VAL`, `HUGE_VALF`, or `HUGE_VALL` is returned (according to the return type and sign of the value),

Secondly, the way I read the spec, in particular the continuation of the sentence above:
> and the value of the macro ERANGE is stored in errno

the return of ±`HUGE_VAL` is simply the most reasonable return value, and one that the application code may be able to work with if it doesn't check for an overflow, but for detecting an overflow it only `errno` matters:
```C
errno=0;
dbl=strtod(nptr,&endptr);
if(errno==ERANGE) {
  // overflow occurred
  assert(dbl == HUGE_VAL || dbl == -HUGE_VAL);
}
```

From the spec there's another error case that seems to be handled incorrectly by this code:
> 7.20.1.3:
> (¶7) If the subject sequence is empty or does not have the expected form, no conversion is performed; the value of nptr is stored in the object pointed to by endptr, provided that endptr is not a null pointer.
> (¶10 If no conversion could be performed, zero is returned.

Obviously, zero is a completely expected return value, so I the way I read it, `endptr==nptr || errno==ERANGE` is the sufficient and necessary condition for detecting an invalid conversion.

The cases of underflow I think we can ignore, and for us it probably makes sense to reject NaNs. I checked the current uses and they are few: `ddsrt_strtof` (itself unused) and something in my crazy little [oneliner](https://github.com/eclipse-cyclonedds/cyclonedds/blob/8638e1fa1e8ae9f964c91ad4763653702b8f91e0/src/core/ddsc/tests/test_oneliner.c#L398). Maybe we should use it instead of `sscanf` more often, but I don't think we'll ever have to worry about underflows, and NaNs would probably only cause trouble, so rejecting them (by default, if need be) makes sense.

If we make the effort of fixing some part of it, let's really fix it and make sure the documentation is correct 😀

Reply:
My reading of `strtod`'s spec was the same as yours and so I've changed it to using `errno==ERANGE` and essentially `endptr==nptr` to check that the parse was good. I also added an `isnan` and an `isinf` check as a condition for resulting in a `DDS_RETCODE_OUT_OF_RANGE` + some documentation to that effect.